shared/typetracking/codeql/typetracking/TypeTracking.qll

```diff
@@ -130,499 +301,510 @@ private module Cached {
+      exists(StepSummary summary |
+        smallStep(_, pragma[only_bind_out](nodeTo), pragma[only_bind_into](summary)) and
+        result = pragma[only_bind_into](pragma[only_bind_out](this)).prepend(summary) and
+        smallStep(nodeFrom, pragma[only_bind_into](pragma[only_bind_out](nodeTo)), summary)
+      )
+      or
+      simpleLocalSmallStep(nodeFrom, nodeTo) and
+      result = this
+    }
+
+    /**
+     * Gets a forwards summary that is compatible with this backwards summary.
+     * That is, if this summary describes the steps needed to back-track a value
+     * from `sink` to `mid`, and the result is a valid summary of the steps needed
+     * to track a value from `source` to `mid`, then the value from `source` may
+     * also flow to `sink`.
+     */
+    TypeTracker getACompatibleTypeTracker() {
+      exists(boolean hasCall | result = MkTypeTracker(hasCall, content) |
+        hasCall = false or hasReturn = false
+      )
+    }
   }
 
-  /**
-   * Gets a forwards summary that is compatible with this backwards summary.
-   * That is, if this summary describes the steps needed to back-track a value
-   * from `sink` to `mid`, and the result is a valid summary of the steps needed
-   * to track a value from `source` to `mid`, then the value from `source` may
-   * also flow to `sink`.
-   */
-  TypeTracker getACompatibleTypeTracker() {
-    exists(boolean hasCall | result = MkTypeTracker(hasCall, content) |
-      hasCall = false or this.hasReturn() = false
-    )
+  /** Provides predicates for implementing custom `TypeBackTracker`s. */
+  module TypeBackTracker {
+    /**
+     * Gets a valid end point of type back-tracking.
+     */
+    TypeBackTracker end() { result.end() }
   }
-}
 
-/** Provides predicates for implementing custom `TypeBackTracker`s. */
-module TypeBackTracker {
+  signature predicate endpoint(Node node);
+
   /**
-   * Gets a valid end point of type back-tracking.
+   * Given a source definition, constructs the default forward type tracking from
+   * those sources.
    */
-  TypeBackTracker end() { result.end() }
+  module TypeTrack<endpoint/1 source> {
+    pragma[assume_small_delta]
+    private Node flow(TypeTracker tt) {
+      tt.start() and source(result)
+      or
+      exists(TypeTracker ttMid | tt = ttMid.step(flow(ttMid), result))
+    }
+
+    /**
+     * Holds if the given source flows to `n`.
+     */
+    predicate flowsTo(Node n) { flowsTo(flow(TypeTracker::end()), n) }
+
+    /**
+     * Given a sink definition, constructs the relation of edges that can be used
+     * in a source-sink path and calculates the set of source-sink pairs.
+     */
+    module Graph<endpoint/1 sink> {
+      private newtype TPathNode = MkPathNode(Node node, TypeTracker tt) { node = flow(tt) }
+
+      /**
+       * A node on a path that is reachable from a source. This is a pair of a
+       * `Node` and a `TypeTracker`.
+       */
+      class PathNode extends TPathNode {
```

Comment:
I think this exposed class should only contain reverse-pruned nodes, just like in normal data flow.